### ISSUE TITLE
[Question] Best way to filter like startsWith / wildcard in the end?

### ISSUE THREAD

**Issue by pwFoo** [leegeunhyeok/bxd#5]
Hi,
great project! Thanks for sharing that!

I see two ways to filter in find method. 
1. range
2. JS callback

Is there a performance difference between? I don't know how range or callback works. Feels like range will filter during query the indexeddb and to use callback the hole database or pre-filtered by range entries will retrieved?

Long question short. What is the best way to filter idb like `startsWith()` with performance in mind?

**Comment by leegeunhyeok** [comment]
Hi, @pwFoo!

**Performance and Limitations**
In general, range is faster when performing a range search, because range search is only available on indexed fields.

However, range doesn't provide a way to do anything other than exact matching or range (inequality) traversal. ([IDBKeyRange](https://developer.mozilla.org/en-US/docs/Web/API/IDBKeyRange)) So, It doesn't supports conditions like startsWith or wildcard match.

**How it works**
1. Range: Reflected immediately in the query and processed internally by IndexedDB.
2. JS callback: It works on the basis of an [IDBCursor](https://developer.mozilla.org/en-US/docs/Web/API/IDBCursor) and sequentially traverses the records through the cursor.

---

The bottom line is that if you want to implement a filter like startsWith or wildcard match, you should do it on a callback basis.

**Comment by pwFoo** [comment]
Hi @leegeunhyeok,

The JS Callback version is working with cursor + bound and executed for each idb entry? 
https://developer.mozilla.org/en-US/docs/Web/API/IDBKeyRange/bound_static

Found a example with `\uffff`:
```
IDBKeyRange.bound(term, `${term}\uffff`, true, false)
```

Performance shouldn't a problem for now because it should be a small idb table... Just to use a recommened solution here :)

**Comment by leegeunhyeok** [comment]
Oh, Looks like a nice trick!

You can use both range and js callback filter like this.

```js
// 1. Query only records where `data` field starts with `${value}` and bring them to the cursor
// 2. For each record result from step 1, call a js callback to filter.
MyBox.find(
  {
    value: BoxDB.Range.bound(value, `${value}\uffff`),
    index: 'data',
  },
  (data) => /* custom filter */,
  (data) => /* custom filter */,
);
```

**Comment by pwFoo** [comment]
Second question...

How to use autoincrement key?
Looks like I can't omit the key field in `add()` method?
```
User = db.create('user', {
    name: { type: BoxDB.Types.STRING, index: true },
    age: { type: BoxDB.Types.NUMBER, index: true },
    id: { type: BoxDB.Types.NUMBER, key: true, autoIncrement: true }
})
```

Tried to omit...
```
User.add({ name: "My Name", age: 100 })
bxd.min.js:1 Uncaught t: Data is not valid
```

**Comment by leegeunhyeok** [comment]
The `autoIncrement` is object store(box)'s option.

you should pass it to third parameter like this.

```diff
User = db.create('user', {
    name: { type: BoxDB.Types.STRING, index: true },
    age: { type: BoxDB.Types.NUMBER, index: true },
-   id: { type: BoxDB.Types.NUMBER, key: true, autoIncrement: true }
+   id: { type: BoxDB.Types.NUMBER, key: true }
- });
+ }, { autoIncrement: true });
```

for more details, https://github.com/leegeunhyeok/bxd/issues/6#issuecomment-1806471372

**Comment by pwFoo** [comment]
Thanks!

**Comment by pwFoo** [comment]
@leegeunhyeok Example should be added to documentation :)
